cpp/examples/parquet/parquet-stream-api/stream-reader-writer.cc

```diff
@@ -0,0 +1,213 @@
+      parquet::ConvertedType::INT_8));
+
+  fields.push_back(parquet::schema::PrimitiveNode::Make(
+      "uint16_field", parquet::Repetition::REQUIRED, parquet::Type::INT32,
+      parquet::ConvertedType::UINT_16));
+
+  fields.push_back(parquet::schema::PrimitiveNode::Make(
+      "int32_field", parquet::Repetition::REQUIRED, parquet::Type::INT32,
+      parquet::ConvertedType::INT_32));
+
+  fields.push_back(parquet::schema::PrimitiveNode::Make(
+      "uint64_field", parquet::Repetition::REQUIRED, parquet::Type::INT64,
+      parquet::ConvertedType::UINT_64));
+
+  fields.push_back(parquet::schema::PrimitiveNode::Make(
+      "double_field", parquet::Repetition::REQUIRED, parquet::Type::DOUBLE,
+      parquet::ConvertedType::NONE));
+
+  // User defined timestamp type.
+  fields.push_back(parquet::schema::PrimitiveNode::Make(
+      "timestamp_field", parquet::Repetition::REQUIRED, parquet::Type::INT64,
+      parquet::ConvertedType::TIMESTAMP_MICROS));
+
+  return std::static_pointer_cast<parquet::schema::GroupNode>(
+      parquet::schema::GroupNode::Make("schema", parquet::Repetition::REQUIRED, fields));
+}
+
+struct TestData {
+  static const int num_rows = 2000;
+
+  static std::string getString(const int i) { return "Str #" + std::to_string(i); }
+  static char getChar(const int i) { return i & 1 ? 'M' : 'F'; }
+  static std::int8_t getInt8(const int i) {
+    return static_cast<std::int8_t>((i % 256) - 128);
+  }
+  static std::uint16_t getUInt16(const int i) { return static_cast<std::uint16_t>(i); }
+  static std::int32_t getInt32(const int i) { return 3 * i - 17; }
+  static std::uint64_t getUInt64(const int i) { return (1ul << 40) + i * i + 101; }
+  static double getDouble(const int i) { return 6.62607004e-34 * 3e8 * i; }
+  static UserTimestamp getUserTimestamp(const int i) {
+    return UserTimestamp{std::chrono::microseconds{(ts_offset + 3 * i) * 1000000}};
+  }
+
+  static char char4_array[4];
+  static std::time_t ts_offset;
+};
+
+char TestData::char4_array[] = "XYZ";
+std::time_t TestData::ts_offset;
+
+void write_parquet_file() {
+  std::shared_ptr<arrow::io::FileOutputStream> outfile;
+
+  PARQUET_THROW_NOT_OK(
+      arrow::io::FileOutputStream::Open("parquet-stream-api-example.parquet", &outfile));
+
+  parquet::WriterProperties::Builder builder;
+
+  builder.compression(parquet::Compression::BROTLI);
+
+  auto file_writer{
+      parquet::ParquetFileWriter::Open(outfile, get_schema(), builder.build())};
+
+  parquet::StreamWriter os{file_writer};
+
+  os.SetMaxRowGroupSize(1000);
+
+  for (auto i = 0; i < TestData::num_rows; ++i) {
+    os << TestData::getString(i);
+    os << TestData::getChar(i);
+    os << TestData::char4_array;
+    os << TestData::getInt8(i);
+    os << TestData::getUInt16(i);
+    os << TestData::getInt32(i);
+    os << TestData::getUInt64(i);
+    os << TestData::getDouble(i);
+    os << TestData::getUserTimestamp(i);
+    os.nextRow();
```

Comment:
it might be nice to provide an additional singleton class EndRow and EndRowGroup that can be inserted into the stream directly to make the api more consistent.

Reply:
Good idea!